dali/util/local_file.cc

```diff
@@ -15,54 +15,178 @@
 
 #include <errno.h>
 #include <sys/stat.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <linux/sysctl.h>
+#include <sys/syscall.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <unistd.h>
 #include <string>
 #include <cstdio>
 #include <cstring>
-
-#ifdef _MSC_VER
-#include <Windows.h>
-#define stat _stat64
-#endif
+#include <map>
+#include <mutex>
+#include <algorithm>
+#include <tuple>
 
 #include "dali/util/local_file.h"
 #include "dali/error_handling.h"
 
+static int _sysctl(struct __sysctl_args *args);
+
+static int get_max_vm_cnt() {
+  struct __sysctl_args args = {0, };
+  int vm_cnt;
+  size_t vm_cnt_sz;
+  int name[] = { CTL_VM, VM_MAX_MAP_COUNT };
+
+  args.name = name;
+  args.nlen = sizeof(name)/sizeof(name[0]);
+  args.oldval = &vm_cnt;
+  args.oldlenp = &vm_cnt_sz;
+
+  vm_cnt_sz = sizeof(vm_cnt);
+
+  if (syscall(SYS__sysctl, &args) == -1) {
+    // fallback to reading /proc
+    FILE * fp;
+    int constexpr MAX_BUFF_SIZE = 256;
+    char buffer[MAX_BUFF_SIZE] = {0, };
+    fp = std::fopen("/proc/sys/vm/max_map_count", "r");
+    vm_cnt_sz = std::fread(buffer, 1, MAX_BUFF_SIZE, fp);
+    vm_cnt = std::stoi(buffer, nullptr);
+  }
+  return vm_cnt;
+}
+
+static void *file_map(const char *path, size_t *length) {
+  int fd = -1;
+  struct stat s;
+  void *p = nullptr;
+
+  if ((fd = open(path, O_RDONLY)) < 0) {
+    goto fail;
+  }
+
+  if (fstat(fd, &s) < 0) {
+    goto fail;
+  }
+
+  *length = (size_t)s.st_size;
+  if ((p = mmap(nullptr, *length, PROT_READ, MAP_PRIVATE, fd, 0)) == MAP_FAILED) {
+    p = nullptr;
+    goto fail;
+  }
+
+fail:
+  if (p == nullptr) {
+    DALI_FAIL("File mapping failed: " + path);
+  }
+  if (fd > -1) {
+    close(fd);
+  }
+  return p;
+}
+
 namespace dali {
 
+typedef std::tuple<void*, size_t> mapped_file;
+typedef std::tuple<std::string, size_t> mapped_file_rev;
+
+// limit to half of allowed mmaped files
+static const unsigned int DALI_MAX_MV_CNT = get_max_vm_cnt() / 2;
```

Comment:
UPPERCASE_IDENTIFIER suggests a compile-time constant or even a `#define`, not a _variable_ (albeit write-once).